File: python/recipe-printer-server/RGB.py
#!/usr/bin/env python2

import math

class RGB:
    
    r = 0
    g = 0
    b = 0
    
    convMatrix = [
        1, 0, 1.13983,
        1, -0.39465, -0.58060,
        1, 2.03211, -0
    ]
    
    def __init__(self, r=0, g=0, b=0):
        self.r = r / 255.0
        self.g = g / 255.0
        self.b = b / 255.0
    
    @classmethod
    def fromFloat(cls, r, g, b):
        return cls(r * 255, g * 255, b * 255)
    
    @classmethod
    def fromYUV(cls, yuv=None):
        if yuv == None:
            return cls()
        else:
            r = min((yuv.y + 1.140 * yuv.v), 1);
            g = min((yuv.y - 0.395 * yuv.u - 0.581 * yuv.v), 1);
            b = min((yuv.y + 2.032 * yuv.u), 1);
            return cls.fromFloat(r, g, b)
        
    @classmethod
    def fromFloatYUV(cls, y=0, u=0, v=0):
        r = min((y + 1.140 * v), 1);
        g = min((y - 0.395 * u - 0.581 * v), 1);
        b = min((y + 2.032 * u), 1);
        return cls.fromFloat(r, g, b)

    #def setcolorHEX(self, hexString):
    #    if len(hexString) == 6:
    #        self.r = int(hexString[0:2], 16) / 255.0
    #        self.g = int(hexString[2,4], 16) / 255.0
    #        self.b = int(hexString[4:], 16) / 255.0
    #    else:
    #        print("RGB: not a valid hex value")
    
    #def setcolorYUVo(self, yuv):
    #    self.r = math.min((yuv.y + 1.140 * yuv.v), 1);
    #    self.g = math.min((yuv.y - 0.395 * yuv.u - 0.581 * yuv.v), 1);
    #    self.b = math.min((yuv.y + 2.032 * yuv.u), 1);
        
    #def setHEX(self, hexString):
    #    if len(hexString) == 6:
    #        self.r = int(hexString[0:2], 16) / 255.0
    #        self.g = int(hexString[2,4], 16) / 255.0
    #        self.b = int(hexString[4:], 16) / 255.0
    #    else:
    #        print("RGB: not a valid hex value")
        
    def toString(self):
        return "["+str(self.r * 255.0)+","+str(self.g * 255.0)+","+str(self.b * 255.0)+"]";

File: processing/applet/RGB.java
package server;


/**
 * Color stored in RGB space. Colors are stored as values from [0 - 1]
 * 
 * @author Christopher Gebhardt
 * @date 04.06.2014
 *
 */
public class RGB {

	public float r;
	public float g;
	public float b;

	/**
	 * conversion matrix for yuv to rgb.
	 * [conversionMatrix] * (yuv)
	 */
	float[] convMatrix = new float[]{
			1f, 0f, 1.13983f,
			1f, -0.39465f, -0.58060f,
			1f, 2.03211f, -0f
	};

	public RGB(float red, float green, float blue){
		r = red;
		g = green;
		b = blue;
	}


	public RGB(int red, int green, int blue){
		r = red / 255f;
		g = green / 255f;
		b = blue / 255f;
	}

	public RGB(String hexString){
		if(hexString.length() == 6){
			r = Integer.parseInt(hexString.substring(0,2), 16) / 255f;
			g = Integer.parseInt(hexString.substring(2,4), 16) / 255f;
			b = Integer.parseInt(hexString.substring(4), 16) / 255f;
		} else {
			System.out.println("RGB: not a valid hex value "+hexString);
		}
	}

	public RGB(YUV yuv){
//		r = convMatrix[0] * yuv.y + convMatrix[1] * yuv.u + convMatrix[2] * yuv.v;
//		g = convMatrix[3] * yuv.y + convMatrix[4] * yuv.u + convMatrix[5] * yuv.v;
//		b = convMatrix[6] * yuv.y + convMatrix[7] * yuv.u + convMatrix[8] * yuv.v;
//		r = (yuv.y + 1.140f * yuv.v) / 255f;
//		g = (yuv.y - 0.395f * yuv.u - 0.581f * yuv.v) / 255f;
//		b = (yuv.y + 2.032f * yuv.u) / 255f;
		
		r = Math.min((yuv.y + 1.140f * yuv.v), 1f);
		g = Math.min((yuv.y - 0.395f * yuv.u - 0.581f * yuv.v), 1f);
		b = Math.min((yuv.y + 2.032f * yuv.u), 1f);
	}

	public void set(String hexString){
		if(hexString.length() == 6){
			try{
				r = Integer.parseInt(hexString.substring(0,2), 16) / 255f;
				g = Integer.parseInt(hexString.substring(2,4), 16) / 255f;
				b = Integer.parseInt(hexString.substring(4), 16) / 255f;
			} catch(NumberFormatException e){
				System.out.println("RGB: not a valid hex value " + hexString);
			}
		} else {
			System.out.println("RGB: not a valid hex value " + hexString);
		}
	}

	public String toString(){
		return "["+(int)(r * 255f)+","+(int)(g * 255f)+","+(int)(b * 255f)+"]";
	}
}